### PR TITLE
Allow setting merge method through labels

### PR DESCRIPTION
**Pull request by michaelbeaumont**
Hi, this PR is meant also as a jumping off point for discussion around the idea. I'm not that attached to the exact names but using labels like this seems like a ergonomic, backwards compatible way to support changing the merge method for specific PRs.

### CODE REVIEW

**Comment by karfau** [comment]
Something like what the title suggest would be awesome!
Currently our "workaround" is 
- Using https://github.com/bettermarks/gh-find-current-pr to set `label_<name>` to `true`
- having conditional runs of the automerge-action with all the same settings beside the merge method

```
jobs:
  detect:
    runs-on: ubuntu-latest
    steps:
      - id: detectPR
        uses: "bettermarks/gh-find-current-pr@bettermarks"
        with:
          github-token: ${{ secrets.GITHUB_TOKEN }}
      - name: Automerge
        if: success() && steps.detectPR.outputs.label_automerge
        uses: "pascalgn/automerge-action@master"
        env:
          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
          MERGE_DELETE_BRANCH: "true"
      - name: Autorebase
        if: success() && steps.detectPR.outputs.label_autorebase
        uses: "pascalgn/automerge-action@master"
        env:
          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
          MERGE_LABELS: "autorebase"
          MERGE_METHOD: "rebase"
      - name: Autosquash
        if: success() && steps.detectPR.outputs.label_autosquash
        uses: "pascalgn/automerge-action@master"
        env:
          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
          MERGE_LABELS: "autosquash"
          MERGE_METHOD: "squash"
```
Which also means that the mentioned gh-find-current-pr action needs to have all the same logic finding the correct PR based on the events, which is tricky, not only from a maintenance perspective.

Personally I would still prefer that it's a new feature that is enabled via an option,
- so backward compatibility is not a concern
- it would be possible to allow just setting one label for that feature (I would expect most people only using one "required" label), like in my example above: `automerge`, `autosquash`, `autorebase`.

**Comment by michaelbeaumont** [comment]
1) A flag makes sense
2) Can you clarify what you mean by one label? With this PR one would only need apply e.g. `automerge.squash` or `<required>.<type>`

**Comment by karfau** [comment]
Just checked your code again and realized that I misunderstood it when I first read it.
I understood that the `.method` label would be required in addition.
But that was my mistake in reading the code.
So just ignore that part.

**Comment by pascalgn** [comment]
Hey, sorry for the delay!

I think this is a nice feature, but I know that there are several people who use highly customized configuration values, so forcing the fixed merge/rebase/squash suffixes feels wrong to me.

Instead, my suggestion would be the following, although maybe not ideal, either:

1. change `MERGE_LABELS` option to support `?` prefix, meaning when using it like `MERGE_LABELS: "done,?automerge/autorebase/autosquash,!wip"` it would mean that a PR is only merged if 1. the _done_ label is present, 2. the _wip_ label is _not_ present and 3. _exactly one_ of the automerge, autorebase _or_ autosquash labels is present
2. change `MERGE_METHOD` to support the configuration like `automerge=merge,autorebase=rebase,autosquash=squash`

I think with these 2 changes, we will keep the flexibility of the action, while still achieving the goal of this PR. Unfortunately, I think this is a bit more effort than the current PR, but I think it would be worth it.

Let me know what you think!

**Comment by michaelbeaumont** [comment]
@pascalgn 

2. What about leaving `MERGE_METHOD` as is (before the PR) and adding `MERGE_METHOD_LABELS` with the format you suggested? Then we have `MERGE_METHOD` be the default if `MERGE_METHOD_LABELS` isn't present.
1. What if we have `MERGE_METHOD_LABEL_REQUIRED=(true|false)` instead of forcing the user to write `?automerge/autorebase/autosquash`? Unless you want the "exactly one of" option in general.

**Comment by TGTGamer** [comment]
Just wanted to throw my hat in the loop, would love to see this. 

MY current work around is to use env variables which are set using something like: 

```yml
- run  echo ::set-env name=mergeType::"squash"
```

and when paired with a standard if statement, allows for both a complex and quick solution.

**Comment by pascalgn** [comment]
> 1. What if we have `MERGE_METHOD_LABEL_REQUIRED=(true|false)` instead of forcing the user to write `?automerge/autorebase/autosquash`? Unless you want the "exactly one of" option in general.

Not sure about this one. How would it be possible to customize the labels then, if someone wants to have e.g. merge/rebase/squash as labels?

> 2. What about leaving `MERGE_METHOD` as is (before the PR) and adding `MERGE_METHOD_LABELS` with the format you suggested? Then we have `MERGE_METHOD` be the default if one of the `MERGE_METHOD_LABELS` isn't present.

Yes, that's better than what I suggested! 👍

**Comment by michaelbeaumont** [comment]
@pascalgn Pushed the changes including the 2 points I mentioned.

File: lib/merge.js [hunk 1]
```diff
@@ -255,6 +274,26 @@ function tryMerge(
...
+function getMergeMethod(defaultMergeMethod, mergeMethodLabels, pullRequest) {
+  const foundMergeMethodLabels = pullRequest.labels.flatMap(l =>
+    mergeMethodLabels.filter(ml => ml.label === l.name)
+  );
+  if (foundMergeMethodLabels.length > 0) {
+    const first = foundMergeMethodLabels[0];
+    if (foundMergeMethodLabels.length > 1) {
+      logger.info(
+        `Discovered multiple merge method labels, will merge with method ${first.method}`
+      );
```
Comment (pascalgn) [inline lib/merge.js:286]: This should throw an error instead, IMO

File: lib/common.js [hunk 2]
```diff
@@ -69,6 +69,19 @@ function createConfig(env = {}) {
...
   }
 
+  function parseMergeMethodLabels(str) {
+    if (!str) {
+      return [];
+    }
+    return str.split(",").map(lm => {
+      const [label, method] = lm.split("=");
+      if (!label || !method) {
+        throw new Error(`Couldn't parse ${lm} as label=method`);
```
Comment (pascalgn) [inline lib/common.js:79]:
```suggestion
        throw new Error(`Couldn't parse "${lm}" as "<label>=<method>" expression`);
```

File: README.md [hunk 3]
```diff
@@ -96,6 +96,14 @@ The following merge options are supported:
   or [`squash`](https://help.github.com/en/articles/about-pull-request-merges#squash-and-merge-your-pull-request-commits)
   (squash all commits into a single commit). The default option is `merge`.
 
+- `MERGE_METHOD_LABELS`: Set to allow labels to determine the merge method.
```
Comment (pascalgn) [inline README.md:99]:
```suggestion
- `MERGE_METHOD_LABELS`: Set to allow labels to determine the merge method
  (see `MERGE_METHOD` for possible values).
```

File: README.md [hunk 4]
```diff
@@ -96,6 +96,14 @@ The following merge options are supported:
...
+  For example, `automerge=merge,autosquash=squash,autorebase=rebase`.
+
+  If no such label is present, use `MERGE_METHOD`.
```
Comment (pascalgn) [inline README.md:102]:
```suggestion
  For example, `automerge=merge,autosquash=squash`. If no such label is present, use
  the value from `MERGE_METHOD`.

  The default value is `""`.
```

**Comment by pascalgn** [comment]
Thanks for the PR, it's merged now! 👍

**Pull request merged**